### PR TITLE
[KEYCLOAK-11821] add the account api roles to the client on creation

### PR DESCRIPTION
**Pull request by unly**
Potential fix for https://issues.jboss.org/browse/KEYCLOAK-11821.

### CODE REVIEW

**Comment by stianst** [comment]
This only solves the problem when there is a brand new installation of Keycloak, but doesn't address the problem for existing users that want to migrate their database.

**Review by stianst — CHANGES_REQUESTED** [review]
We need to consider what to do for existing realms. Perhaps the simplest would be to just always add these roles even if the endpoints are preview.

Also, we need some tests to make sure things are added as expected.

**Comment by unly** [comment]
Hey @stianst thanks for getting back. Since the Account API is still under development, I added the required roles just on the creation on a new realm when the feature is enable. Reading your comment I assume you prefer to always have those roles even though users might not use the preview feature at all. So far, I left this out. This would require a migration in the database for existing realms. When you agree on this I can try to provide the scripts and the requested tests.

**Comment by stianst** [comment]
+1 Let's just add the roles always as it simplifies things, and there's no big side-effects of having them there even with the account endpoints disabled.

**Comment by unly** [comment]
Hi @stianst I added the three required roles for the account api to the migration to 8.0.0. Can you please have another look at it

**Comment by stianst** [comment]
@unly afraid 8 is release already, so migration need to go to 9 :(

**Comment by stianst** [comment]
Didn't review the rest, but will do that early next week

File: server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java [hunk 1]
```diff
@@ -55,6 +56,22 @@ public void migrateImport(KeycloakSession session, RealmModel realm, RealmRepres
...
+        RoleModel viewConsentRole = accountClient.addRole(AccountRoles.VIEW_CONSENT);
+        viewConsentRole.setDescription("${role_" + AccountRoles.VIEW_CONSENT + "}");
+        LOG.debugf("Added the role %s to the '%s' client.", AccountRoles.VIEW_CONSENT, Constants.ACCOUNT_MANAGEMENT_CLIENT_ID);
+        RoleModel manageConsentRole = accountClient.addRole(AccountRoles.MANAGE_CONSENT);
+        manageConsentRole.setDescription("${role_" + AccountRoles.MANAGE_CONSENT + "}");
+        LOG.debugf("Added the role %s to the '%s' client.", AccountRoles.MANAGE_CONSENT, Constants.ACCOUNT_MANAGEMENT_CLIENT_ID);
+        manageConsentRole.addCompositeRole(viewConsentRole);
+        LOG.debugf("Added the %s role as a composite role to %s", AccountRoles.VIEW_CONSENT, AccountRoles.MANAGE_CONSENT);
+
+        RoleModel manageAccountRole = accountClient.getRole(AccountRoles.MANAGE_ACCOUNT);
```
Comment (stianst) [inline server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java]: Manage account shouldn't be a composite role. Instead the rest endpoints should check for both manage account and manage consent roles. I believe this was already added

File: server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java [hunk 2]
```diff
@@ -55,6 +56,22 @@ public void migrateImport(KeycloakSession session, RealmModel realm, RealmRepres
 
     protected void migrateRealmCommon(RealmModel realm) {
         addAccountConsoleClient(realm);
+
+        ClientModel accountClient = realm.getClientByClientId(Constants.ACCOUNT_MANAGEMENT_CLIENT_ID);
```
Comment (stianst) [inline server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java]: Extract into a separate method, something like "addAccountRoles(RealmModel realm)"

**Review by stianst — CHANGES_REQUESTED** [review]
Some minor comments. We also need a migration test added.

**Comment by stianst** [comment]
@unly are you able to address the minor comments and add a migration test?

**Comment by unly** [comment]
@stianst thanks for reminding :) i missed it after a couple days off. I added some tests but the travis test still seem to fail with something different

**Review by stianst — APPROVED** [review]

**Comment by stianst** [comment]
@unly failing tests looks completely unrelated, and I can't see how your changes would have any impact on those tests, so they are likely just some unstable tests.

**Comment by stianst** [comment]
Changes look good now. There was an issue in MigrateTo9_0_0 where it invokes addAccountConsoleClient instead of migrateRealmCommon, which results in realms being updated in the DB not having these roles added (see https://github.com/keycloak/keycloak/pull/6694/files#diff-8b9a09a2f90de5a0f095e0e5a0b9db23R49 for the fix).

**Comment by stianst** [comment]
Opened a new PR here https://github.com/keycloak/keycloak/pull/6694 with the fix, so closing this PR.

**Comment by unly** [comment]
Thanks @stianst

**Pull request closed without merging**